### PR TITLE
Exclude the source repo as dependent repo

### PR DESCRIPTION
**Pull request by ericstone93**
Summary of the changes (in less than 80 characters):
Implement the solution suggested in the issue so the source repo is not shown in the top 5 popular repos in the case where the source repo has PackageReference to itself.

* Increase `ReposPerPackage ` from 20 to 21.
* Skip the repo if the Url is the same as the source repo.

Addresses https://github.com/NuGet/NuGetGallery/issues/9441

### CODE REVIEW

**Comment by ericstone93** [comment]
@microsoft-github-policy-service agree

**Comment by ericstone93** [comment]
I tried to run the web locally and noticed I cannot search the `Enums.NET`. Please let me know how I can do local/PPE test. Thanks.

**Comment by joelverhagen** [comment]
@ericstone93, this looks reasonable. As far as testing it locally, this is a bit challenging since you need a "GitHubUsage.v1.json" file locally and some test packages uploaded to your local registry. We don't have an easy way for you to connect directly to our PPE (INT/DEV in our terms) environments because that require DB and blob storage access.

The "GitHubUsage.v1.json" file is produced by this job:
https://github.com/NuGet/NuGetGallery/tree/main/src/NuGet.Jobs.GitHubIndexer
But unfortunately this is one of the jobs where we don't have a good README for OSS contributors to follow and run it locally.

I think whoever on the team reviews your PR can take the responsibility of testing it on our PPE envs.

As a side note, please rebase onto the `dev` branch so this can be merged into our normal development branch (dev) instead of our hotfix branch (main).

I looked the data and tried to find edge cases where a more complex check is needed. These are my suggestions

- Check both the project URL and repository URL (some packages only set repository URL)
- Compare the URLs in a case insensitive way. GitHub accepts any casing for the owner and repository name part of the URL
- Remove the trailing slash before comparing
- Ignore the scheme part of the project or repository URL (I see `git://` URLs in the data)
- Remove a `.git` suffix from the project/repository URL before comparing

I think it makes sense to add a `ComparableGitHubRepository` property to the `DisplayPackageViewModel` model to use instead of `Model.RepositoryUrl` in the `.cshtml`. This could new property take the repository or project URL that has a `github.com` URI authority (maybe preferring the repository URL). This also allows you to do the normalization in a `.cs` file which is easier to unit test.

You could put the logic here: https://github.com/NuGet/NuGetGallery/blob/main/src/NuGetGallery/Helpers/ViewModelExtensions/DisplayPackageViewModelFactory.cs

You could put the unit tests here:
https://github.com/NuGet/NuGetGallery/blob/main/tests/NuGetGallery.Facts/ViewModels/DisplayPackageViewModelFacts.cs

Here are some top examples of edge cases.:

<details>

Id | ProjectUrl | RepositoryUrl | DependentGitHubRepository | Normalization
-- | -- | -- | -- | --
Microsoft.SqlServer.Server | https://aka.ms/sqlclientproject | https://github.com/dotnet/sqlclient | dotnet/SqlClient | Case insensitive compare
MediatR.Contracts |   | https://github.com/jbogard/MediatR.git | jbogard/MediatR | Remove .git suffix
Xamarin.Forms | http://xamarin.com/forms | https://github.com/xamarin/xamarin.forms | xamarin/Xamarin.Forms | Case insensitive compare
Datadog.Trace | https://docs.datadoghq.com/tracing/trace_collection/dd_libraries | https://github.com/DataDog/dd-trace-dotnet.git | DataDog/dd-trace-dotnet | Remove .git suffix
App.Metrics.Abstractions | https://app-metrics.io/ | git://github.com/AppMetrics/AppMetrics | AppMetrics/AppMetrics | Ignore git:// scheme
App.Metrics.Formatters.Ascii | https://app-metrics.io/ | git://github.com/AppMetrics/AppMetrics | AppMetrics/AppMetrics | Ignore git:// scheme
App.Metrics.Core | https://app-metrics.io/ | git://github.com/AppMetrics/AppMetrics | AppMetrics/AppMetrics | Ignore git:// scheme
App.Metrics.Concurrency | https://app-metrics.io/ | git://github.com/AppMetrics/AppMetrics | AppMetrics/AppMetrics | Ignore git:// scheme
App.Metrics.Formatters.Json | https://app-metrics.io/ | git://github.com/AppMetrics/AppMetrics | AppMetrics/AppMetrics | Ignore git:// scheme
App.Metrics | https://app-metrics.io/ | git://github.com/AppMetrics/AppMetrics | AppMetrics/AppMetrics | Ignore git:// scheme
App.Metrics.Extensions.DependencyInjection | https://app-metrics.io/ | git://github.com/AppMetrics/AppMetrics | AppMetrics/AppMetrics | Ignore git:// scheme
App.Metrics.Extensions.Configuration | https://app-metrics.io/ | git://github.com/AppMetrics/AppMetrics | AppMetrics/AppMetrics | Ignore git:// scheme
HotChocolate.AspNetCore | https://chillicream.com/ | https://github.com/ChilliCream/graphql-platform.git | ChilliCream/graphql-platform | Remove .git suffix
Microsoft.Azure.DurableTask.AzureStorage | https://github.com/Azure/durabletask/ | https://github.com/Azure/durabletask/ | Azure/durabletask | Remove trailing slash
Microsoft.ApplicationInsights.Kubernetes | https://github.com/Microsoft/ApplicationInsights-Kubernetes/ | https://github.com/Microsoft/ApplicationInsights-Kubernetes/ | microsoft/ApplicationInsights-Kubernetes | Case insensitive compare
App.Metrics.AspNetCore.Hosting | https://app-metrics.io/ | git://github.com/AppMetrics/AppMetrics | AppMetrics/AppMetrics | Ignore git:// scheme
App.Metrics.AspNetCore.Endpoints | https://app-metrics.io/ | git://github.com/AppMetrics/AppMetrics | AppMetrics/AppMetrics | Ignore git:// scheme
App.Metrics.AspNetCore.Tracking | https://app-metrics.io/ | git://github.com/AppMetrics/AppMetrics | AppMetrics/AppMetrics | Ignore git:// scheme
App.Metrics.AspNetCore | https://app-metrics.io/ | git://github.com/AppMetrics/AppMetrics | AppMetrics/AppMetrics | Ignore git:// scheme



</details>

Thanks for taking the time to work on this.

**Comment by erdembayar** [comment]
@ericstone93 

Thank you for your PR. Unfortunately, we are currently very busy and unable to review it at this time. We'll review when we have capacity.

**Comment by erdembayar** [comment]
@ericstone93 
Could you please take a look at comments by Joel? If still no update then we'll close this PR as abandoned.

**Comment by ericstone93** [comment]
> @ericstone93 Could you please take a look at comments by Joel? If still no update then we'll close this PR as abandoned.

Yes, I will take another look this week and submit change.

**Comment by ericstone93** [comment]
I have added applied the suggestion to add `ComparableGitHubRepository`. And would like to clarify if I should compare the URLs in the cshtml? If not, which URL the `ComparableGitHubRepository` should compare with?

File: tests/NuGetGallery.Facts/ViewModels/DisplayPackageViewModelFacts.cs [hunk 1]
```diff
@@ -131,6 +131,30 @@ public void ItDeterminesRepositoryKind(string repoUrl, string repoType, Reposito
             Assert.Equal(expectedUrl, model.RepositoryUrl);
         }
 
+        [Theory]
+        [InlineData("https://github.com/dotnet/sqlclient", "git", "dotnet/SqlClient")]
+        [InlineData("https://github.com/jbogard/MediatR.git", "git", "jbogard/MediatR")]
+        [InlineData("git://github.com/AppMetrics/AppMetrics", null, "AppMetrics/AppMetrics")]
+        [InlineData("https://github.com/Azure/durabletask/", "git", "Azure/durabletask")]
+        [InlineData("https://visualstudio.com", "tfs", "")]
+        public void ItDeterminesComparableGitHubRepository(string repoUrl, string repoType, string ComparableGitHubRepository)
```
Comment (joelverhagen) [inline tests/NuGetGallery.Facts/ViewModels/DisplayPackageViewModelFacts.cs:140]: Nice tests!

File: src/NuGetGallery/ViewModels/DisplayPackageViewModel.cs [hunk 2]
```diff
@@ -138,6 +140,36 @@ public void InitializeRepositoryMetadata(string repositoryUrl, string repository
...
+            if (comparableUrl.StartsWith("git://"))
+            {
+                comparableUrl = comparableUrl.Replace($"git://{githubCom}", "");
+            }
+            else if (comparableUrl.StartsWith("https://"))
+            {
+                comparableUrl = comparableUrl.Replace($"https://{githubCom}", "");
+            }
+
+            ComparableGitHubRepository = comparableUrl;
```
Comment (joelverhagen) [inline src/NuGetGallery/ViewModels/DisplayPackageViewModel.cs:170]: I don't see this new property getting read anywhere, like in the .cshtml.

Comment (joelverhagen) [reply src/NuGetGallery/ViewModels/DisplayPackageViewModel.cs:170]: Ah, I see your comment here now https://github.com/NuGet/NuGetGallery/pull/10347#issuecomment-2744492522. Sorry for missing it. I think the shape of `ComparableGitHubRepository` can match one of the properties on the Repos item, such as the `Id`, e.g. `microsoft/PowerToys` or the `Url` e.g. `https://github.com/microsoft/PowerToys` then you compare against that in the cshtml.

File: src/NuGetGallery/ViewModels/DisplayPackageViewModel.cs [hunk 3]
```diff
@@ -138,6 +140,36 @@ public void InitializeRepositoryMetadata(string repositoryUrl, string repository
             }
         }
 
+        public void InitializeComparableGitHubRepository()
+        {
+            const string githubCom = "github.com/";
+            string comparableUrl = string.Empty;
+
+            if (RepositoryUrl.Contains(githubCom))
```
Comment (joelverhagen) [inline src/NuGetGallery/ViewModels/DisplayPackageViewModel.cs:148]: This needs to handle null Repository URL. It's failing on UTs:
```

  Stack Trace:
     at NuGetGallery.DisplayPackageViewModel.InitializeComparableGitHubRepository() in D:\a\_work\1\ng\src\NuGetGallery\ViewModels\DisplayPackageViewModel.cs:line 146
   at NuGetGallery.DisplayPackageViewModelFactory.SetupCommon(DisplayPackageViewModel viewModel, Package package, String pushedBy, IReadOnlyDictionary`2 packageKeyToDeprecation, IReadOnlyDictionary`2 packageKeyToVulnerabilities) in D:\a\_work\1\ng\src\NuGetGallery\Helpers\ViewModelExtensions\DisplayPackageViewModelFactory.cs:line 171
   at NuGetGallery.PackagesController.<DisplayPackage>d__66.MoveNext() in D:\a\_work\1\ng\src\NuGetGallery\Controllers\PackagesController.cs:line 965
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at NuGetGallery.PackagesControllerFacts.TheDisplayPackageMethod.<GetActionResultForPackageStatusAsync>d__18.MoveNext() in D:\a\_work\1\ng\tests\NuGetGallery.Facts\Controllers\PackagesControllerFacts.cs:line 745
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at NuGetGallery.PackagesControllerFacts.TheDisplayPackageMethod.<GivenAValidatingPackageThatTheCurrentUserOwnsThenShowIt>d__11.MoveNext() in D:\a\_work\1\ng\tests\NuGetGallery.Facts\Controllers\PackagesControllerFacts.cs:line 596
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
  Failed NuGetGallery.PackagesControllerFacts+TheDisplayPackageMethod.GivenAValidatingPackageThatTheCurrentUserOwnsThenShowIt(packageStatus: FailedValidation) [45 ms]
 ```

Comment (ericstone93) [reply src/NuGetGallery/ViewModels/DisplayPackageViewModel.cs]: Thanks, updated

**Comment by ericstone93** [comment]
I resolve the merge conflict by updating the new number of `ReposPerPackage` 20 to 21 due to the possible self-duplicate case in this PR.

File: src/NuGetGallery/ViewModels/DisplayPackageViewModel.cs [hunk 4]
```diff
@@ -138,6 +140,36 @@ public void InitializeRepositoryMetadata(string repositoryUrl, string repository
             }
         }
 
+        public void InitializeComparableGitHubRepository()
+        {
+            const string githubCom = "github.com/";
+            string comparableUrl = string.Empty;
+
+            if (RepositoryUrl !=null && RepositoryUrl.Contains(githubCom))
```
Comment (joelverhagen) [inline src/NuGetGallery/ViewModels/DisplayPackageViewModel.cs:148]:
```suggestion
            if (RepositoryUrl != null && RepositoryUrl.Contains(githubCom))
```
nit

File: tests/NuGetGallery.Facts/ViewModels/DisplayPackageViewModelFacts.cs [hunk 5]
```diff
@@ -131,6 +131,30 @@ public void ItDeterminesRepositoryKind(string repoUrl, string repoType, Reposito
...
+                RepositoryType = repoType,
+                PackageRegistration = new PackageRegistration
+                {
+                    Owners = Enumerable.Empty<User>().ToList(),
+                    Packages = Enumerable.Empty<Package>().ToList()
+                }
+            };
+
+            var model = CreateDisplayPackageViewModel(package, currentUser: null, packageKeyToDeprecation: null, readmeHtml: null);
+            Assert.Equal(ComparableGitHubRepository.ToLowerInvariant(), model.ComparableGitHubRepository);
```
Comment (joelverhagen) [inline tests/NuGetGallery.Facts/ViewModels/DisplayPackageViewModelFacts.cs:155]:
```suggestion
            Assert.Equal(ComparableGitHubRepository, model.ComparableGitHubRepository, StringComparer.OrdinalIgnoreCase);
```

File: src/NuGetGallery/Views/Packages/DisplayPackage.cshtml [hunk 6]
```diff
@@ -920,8 +920,21 @@
...
                                         </thead>
                                         <tbody class="no-border">
+                                            @{ var selectedItemsCount = 0; }
                                             @foreach (var item in Model.GitHubDependenciesInformation.Repos.Select((elem, i) => new { Value = elem, Idx = i }))
                                             {
+                                                if (string.Equals(item.Value.id, Model.ComparableGitHubRepository, StringComparison.OrdinalIgnoreCase))
+                                                {
+                                                    continue;
+                                                }
+                                                else if (selectedItemsCount == 5)
```
Comment (joelverhagen) [inline src/NuGetGallery/Views/Packages/DisplayPackage.cshtml:930]:
```suggestion
                                                else if (selectedItemsCount == ReposPerPackage.ReposPerPackage - 1)

```

Comment (joelverhagen) [reply src/NuGetGallery/Views/Packages/DisplayPackage.cshtml:930]: We should use the constant instead of having two places to update

File: src/NuGetGallery/Views/Packages/DisplayPackage.cshtml [hunk 7]
```diff
@@ -920,8 +920,21 @@
                                             </tr>
                                         </thead>
                                         <tbody class="no-border">
+                                            @{ var selectedItemsCount = 0; }
                                             @foreach (var item in Model.GitHubDependenciesInformation.Repos.Select((elem, i) => new { Value = elem, Idx = i }))
                                             {
+                                                if (string.Equals(item.Value.id, Model.ComparableGitHubRepository, StringComparison.OrdinalIgnoreCase))
```
Comment (joelverhagen) [inline src/NuGetGallery/Views/Packages/DisplayPackage.cshtml:926]:
```suggestion
                                                if (string.Equals(item.Value.Id, Model.ComparableGitHubRepository, StringComparison.OrdinalIgnoreCase))
```

this is failing compile on CI:
```
         d:\a\_work\1\ng\src\NuGetGallery\Views\Packages\DisplayPackage.cshtml(926): error CS1061: 'NuGetGallery.RepositoryInformation' does not contain a definition for 'id' and no extension method 'id' accepting a first argument of type 'NuGetGallery.RepositoryInformation' could be found (are you missing a using directive or an assembly reference?) [D:\a\_work\1\ng\src\NuGetGallery\NuGetGallery.csproj]
```

**Comment by joelverhagen** [comment]
@ericstone93 - thanks for applying the feedback. I tested it locally, and noticed the counts (total count in the heading, and list size in the description) were off by one when excluding. I updated the PR to fix this. Once the CI passes, we should be good to merge. Thank you!

**Review by joelverhagen — APPROVED** [review]

**Comment by joelverhagen** [comment]
Merged. Thanks so much @ericstone93!

**Comment by ericstone93** [comment]
Thanks for the help on the fix and merge the PR @joelverhagen
